src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -929,7 +935,19 @@ HeapWord* ShenandoahHeap::allocate_memory_under_lock(ShenandoahAllocRequest& req
 
 HeapWord* ShenandoahHeap::mem_allocate(size_t size,
                                         bool*  gc_overhead_limit_was_exceeded) {
+  *gc_overhead_limit_was_exceeded = false;
   ShenandoahAllocRequest req = ShenandoahAllocRequest::for_shared(size);
+  if (get_gc_no_progress_count() > ShenandoahNoProgressThreshold) {
```

Comment:
Move this block to `ShenandoahHeap::allocate_memory`? This would require adding a (default) argument to that method, but it is a good tradeoff for putting all mechanics around this into a single method.

Reply:
I put it here to keep it on the slow(er) path so that allocations for LABs won't have additional checks on the progress counter. Also, a failed LAB allocation will come back to `mem_allocate` to try for a shared allocation.